src/components/tooltip/tooltip.js

```diff
@@ -0,0 +1,212 @@
+      childrenProps.onMouseOver(event);
+    }
+
+    if (event.type === 'focus' && childrenProps.onFocus) {
+      childrenProps.onFocus(event);
+    }
+
+    if (!this.state.open && !this.isControlled) {
+      this.setState({
+        open: true,
+      });
+    }
+
+    if (this.props.onOpen) {
+      this.props.onOpen(event);
+    }
+  };
+
+  handleLeave = event => {
+    const childrenProps = this.props.children.props;
+    clearTimeout(this.leaveTimer);
+
+    if (event.type === 'mouseleave' && childrenProps.onMouseLeave) {
+      childrenProps.onMouseLeave(event);
+    }
+
+    if (event.type === 'blur' && childrenProps.onBlur) {
+      childrenProps.onBlur(event);
+    }
+
+    if (this.props.leaveDelay) {
+      this.leaveTimer = setTimeout(() => {
+        this.handleClose();
+      }, this.props.leaveDelay);
+    } else {
+      this.handleClose();
+    }
+  };
+
+  handleClose = event => {
+    if (!this.isControlled) {
+      this.setState({
+        open: false,
+      });
+    }
+    if (this.props.onClose) {
+      this.props.onClose(event);
+    }
+  };
+
+  setChildrenRef = ref => {
+    this.childrenRef = ref;
+  };
+
+  isControlled = this.props.isOpen != null;
+
+  render() {
+    const open = this.isControlled ? this.props.isOpen : this.state.open;
+
+    const childrenProps = {
+      'aria-describedby': open ? this.state.id : null,
+      // for seo and accessibility, we add the tooltip's title
+      // as a native title when the title is hidden
+      title:
+        !open && typeof this.props.title === 'string' ? this.props.title : null,
+      ...this.props.children.props,
+      onMouseOver: this.handleEnter,
+      onMouseLeave: this.handleLeave,
+      onFocus: this.handleEnter,
+      onBlur: this.handleLeave,
+    };
+
+    return (
+      <Manager>
+        <Reference innerRef={this.setChildrenRef}>
+          {({ ref }) => (
+            <RootRef rootRef={ref}>
+              {React.cloneElement(this.props.children, {
+                ...childrenProps,
+              })}
+            </RootRef>
+          )}
+        </Reference>
+        {open && (
+          <Popper placement={this.props.placement}>
+            {({ ref, style, placement }) => (
+              <div
+                ref={ref}
+                css={{
+                  ...style,
+                  ...getBodyStyles({
+                    constraint: this.props.horizontalConstraint,
+                    placement,
+                  }),
+                }}
+                data-placement={placement}
+              >
+                {this.props.title}
+              </div>
+            )}
+          </Popper>
+        )}
+      </Manager>
+    );
+  }
+}
+
+Tooltip.propTypes = {
+  children: PropTypes.node.isRequired,
+  horizontalConstraint: PropTypes.oneOf(['xs', 's', 'm', 'l', 'xl', 'scale'])
+    .isRequired,
+  leaveDelay: PropTypes.number.isRequired,
+  isOpen: PropTypes.bool,
+  onClose: PropTypes.func,
```

Comment:
would a `onClose: requiredIf(PropTypes.func, props => !props.isOpen)` make sense?

Reply:
Just to understand, by `!props.isOpen`, do you mean that the component is controlled by it's internal state, or it's controlled by `isOpen` as a prop?

Reply:
Maybe it doesn't make sense true. I've just seen this elsewhere. If you pass e.g. `isOpen` you mean to control the component. You however can't if don't use the event handlers (which is not true in this case).